Source/WebKit/WebProcess/Plugins/PDF/UnifiedPDF/UnifiedPDFPlugin.mm

```diff
@@ -3284,17 +3284,39 @@ static FloatRect computeMarqueeSelectionRect(const WebCore::FloatPoint& point1,
     unfreezeCursorAfterSelectionDragIfNeeded();
 }
 
-Vector<FloatRect> UnifiedPDFPlugin::boundsForSelection(const PDFSelection *selection, CoordinateSpace inSpace) const
+PDFPageCoverage UnifiedPDFPlugin::pageCoverageForSelection(PDFSelection *selection, FirstPageOnly firstPageOnly) const
 {
     if (!selection || [selection isEmpty])
         return { };
 
-    return makeVector([selection pages], [this, selection, inSpace](PDFPage *page) -> std::optional<FloatRect> {
+    auto pageCoverage = PDFPageCoverage { };
+
+    for (PDFPage *page in [selection pages]) {
         auto pageIndex = m_documentLayout.indexForPage(page);
         if (!pageIndex)
-            return { };
-        return convertUp(CoordinateSpace::PDFPage, inSpace, FloatRect { [selection boundsForPage:page] }, *pageIndex);
-    });
+            continue;
+
+        pageCoverage.append({ *pageIndex, FloatRect { [selection boundsForPage:page] } });
+        if (firstPageOnly == FirstPageOnly::Yes)
+            break;
+    }
+
+    return pageCoverage;
+}
+
+Vector<FloatRect> UnifiedPDFPlugin::boundsForSelection(PDFSelection *selection, CoordinateSpace targetSpace) const
+{
+    auto pageCoverage = pageCoverageForSelection(selection);
+    if (pageCoverage.isEmpty())
+        return { };
+
+    Vector<FloatRect> pageRects;
+    pageRects.reserveInitialCapacity(pageCoverage.size());
+
+    for (auto& page : pageCoverage)
+        pageRects.append(convertUp(CoordinateSpace::PDFPage, targetSpace, page.pageBounds, page.pageIndex));
+
+    return pageRects;
```

Comment:
nit: You could use `Vector::map` for this one? The `map` implementation is basically what you have here!

```suggestion
    return pageCoverageForSelection(selection).map([this, targetSpace](auto& page) {
        return convertUp(CoordinateSpace::PDFPage, targetSpace, page.pageBounds, page.pageIndex);
    };
```